### PR TITLE
extract cluster cidrs process to generic controller manager

### PR DESCRIPTION
**Pull request by m3ngyang**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
There are some exactly duplicated codes for dealing with cluster CIDR, they cloud be extracted to form a general function in generic controller manager.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

### CODE REVIEW

**Comment by jsafrane** [comment]
/ok-to-test

@m3ngyang, are you sure you assigned this to the right guy? I don't know anything about routes.

/unsassign

**Comment by m3ngyang** [comment]
Sorry about the assignment.
I assigned you because you was chosen as the reviewer by the k8s robot.

File: cmd/controller-manager/app/helper.go
Comment (m3ngyang) [reply cmd/controller-manager/app/helper.go]: good, it expresses much more clearly.

File: cmd/controller-manager/app/helper.go
Comment (m3ngyang) [reply cmd/controller-manager/app/helper.go]: ok~

**Review by jpbetz — APPROVED** [review]
/lgtm

Thanks for adding the tests!

**Comment by m3ngyang** [comment]
/assign @andrewsykim
/assign @deads2k
PTAL

**Comment by andrewsykim** [comment]
/assign @khenidak 

re: dual-stack changes

File: cmd/controller-manager/app/helper_test.go
Comment (andrewsykim) [inline cmd/controller-manager/app/helper_test.go]: You can use t.Run to make this a bit cleaner
```
for _, tc := range tcs {
    t.Run(tc.name, func(t *testing.T){ 
         ....
   })
}
```
And that way you don't have to print the test name on failure since t.Run will do that for you

Comment (andrewsykim) [reply cmd/controller-manager/app/helper_test.go]: (no wait group as well)

File: cmd/controller-manager/app/helper_test.go
Comment (andrewsykim) [inline cmd/controller-manager/app/helper_test.go]: What about the net.IPNet values?

File: cmd/controller-manager/app/helper_test.go
Comment (andrewsykim) [inline cmd/controller-manager/app/helper_test.go]: You can simplify this a little bit with embedded structs
```
tests := []struct{
                name       string
		cidrs      string
		ipv6Enable bool
		valid      bool
}{
          {
                     ....
          }
}
```

**Comment by m3ngyang** [comment]
Unit tests have been rewritten according to @andrewsykim 's advice. @jpbetz

**Comment by m3ngyang** [comment]
@andrewsykim @jpbetz this pr needs a `/lgtm` label after rewriting unit tests with `t.Run`.

**Comment by m3ngyang** [comment]
/assign @liggitt 
PTAL~

**Comment by m3ngyang** [comment]
@sttts 
would you like to take a look

**Comment by m3ngyang** [comment]
/assign @liggitt 
would you like to have a look.

**Comment by m3ngyang** [comment]
/assign @liggitt 
do you have any opinion?

**Comment by m3ngyang** [comment]
conflicts solved.
/assign @liggitt

**Comment by m3ngyang** [comment]
/assign @liggitt
PTAL👀

**Comment by m3ngyang** [comment]
@liggitt 
PTAL

**Comment by liggitt** [comment]
this doesn't seem like generic controller manager functionality... the other things built into generic controller manager packages deal with setting up individual clients, enabling/disabling controller loops, ensuring the API server can be contacted, etc. This is a function that the cloud-controller-manager and kube-controller-manager happen to both share, but would be unlikely to be useful to another group of controller loops.

**Comment by m3ngyang** [comment]
I think `cmd/controller-manager` is a common package for kube-controller-manager and cloud-controller-manager to share generic functions, and the only two functions inside the package are called by those two controllers. @liggitt

**Comment by liggitt** [comment]
> I think cmd/controller-manager is a common package for kube-controller-manager and cloud-controller-manager to share generic functions

My point is that this is not a generic function. The types of functions that make sense here are around client-building, enabling/disabling named controller loops, etc.

**Pull request closed without merging**
